presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,283 @@
+                    PlanNodeCostEstimate node2Cost = lookup.getCumulativeCost(node2, session, symbolAllocator.getTypes());
+                    return JoinEnumerator.this.costComparator.compare(session, node1Cost, node2Cost);
+                }
+            };
+            this.costComparator = costComparator;
+        }
+
+        private static boolean isJoinEqualityCondition(Expression expression)
+        {
+            return expression instanceof ComparisonExpression
+                    && ((ComparisonExpression) expression).getType() == EQUAL
+                    && ((ComparisonExpression) expression).getLeft() instanceof SymbolReference
+                    && ((ComparisonExpression) expression).getRight() instanceof SymbolReference;
+        }
+
+        private static JoinNode.EquiJoinClause toEquiJoinClause(ComparisonExpression equality, Set<Symbol> leftSymbols)
+        {
+            boolean alignedComparison = leftSymbols.contains(Symbol.from(equality.getLeft()));
+            Expression leftExpression = alignedComparison ? equality.getLeft() : equality.getRight();
+            Expression rightExpression = alignedComparison ? equality.getRight() : equality.getLeft();
+
+            Symbol leftSymbol = Symbol.from(leftExpression);
+            Symbol rightSymbol = Symbol.from(rightExpression);
+            return new JoinNode.EquiJoinClause(leftSymbol, rightSymbol);
+        }
+
+        /**
+         * This method generates all the ways of dividing  totalNodes into two sets
+         * each containing at least one node. It will generate one set for each
+         * possible partitioning. The other partition is implied in the absent values.
+         * In order not to generate the inverse of any set, we always include the 0th
+         * node in our sets.
+         *
+         * @param totalNodes
+         * @return A set of sets each of which defines a partitioning of totalNodes
+         */
+        public static Stream<Set<Integer>> generatePartitions(int totalNodes)
+        {
+            checkArgument(totalNodes > 0, "totalNodes must be greater than 0");
+            Set<Integer> numbers = IntStream.range(0, totalNodes)
+                    .boxed()
+                    .collect(toImmutableSet());
+            return Sets.powerSet(numbers).stream()
+                    .filter(subSet -> subSet.contains(0))
+                    .filter(subSet -> subSet.size() < numbers.size());
+        }
+
+        public static JoinNode createJoinAccordingToPartitioning(JoinGraphNode joinGraph, Set<Integer> partitioning, PlanNodeIdAllocator idAllocator)
+        {
+            List<PlanNode> sources = joinGraph.getSources();
+            List<PlanNode> leftSources = partitioning.stream()
+                    .map(sources::get)
+                    .collect(toImmutableList());
+            List<PlanNode> rightSources = ImmutableList.copyOf(Sets.difference(ImmutableSet.copyOf(sources), ImmutableSet.copyOf(leftSources)));
```

Comment:
Can't you just use set here? Also `rightSources` order is unpredictable. However, memoization takes order into account